### DISCUSSION TITLE
Am I misunderstanding virtualenvs.options.always-copy?

### DISCUSSION THREAD

**Discussion by allynt** [python-poetry/poetry#7323] [Q&A]
Hello.

I have a python application which uses poetry & docker.  The venv created by poetry is mounted inside my docker container.  

If I create the venv from w/in docker (ie: run `poetry install` as part of the Dockerfile) [**situation 1**] then the ".venv/pyvenv.cfg" file includes content like this:

```
base-executable = /usr/bin/python3
```

And if I create the venv in the local filesystem (ie: run `poetry install` _before_ launching Docker) [**situation 2**] then the ".venv/pyvenv.cfg" file includes content like this:

```
base-executable = /home/myuser/.pyenv/versions/3.10.0/bin/python3.10
```

And sure enough `python` in the venv is symlinked to those files.

Clearly, if I try and run an executable (like, say, `yapf`) from outside docker in **situation 1** it will fail.

I naively assumed that setting `virtualenvs.options.always-copy=true` would actually _copy_ the python executable into the venv instead of symlinking it.  But that had no effect.  

Am I misunderstanding that setting?

**Comment by kler** [comment]
## Hardlinks or Softlinks/Symlinks
I'm not a developer of poetry, but since I'm trying to find out whether poetry is able to create virtual environments using hardlinks or softlinks/symlinks, I stumbled upon this question. As a side note I believe this is an overlooked issue.

I'm, perhaps naively, imagining that poetry could host a global "repository" *(different from the cache, that may be emptied at any time)* that would contain all packages used throughout the system. When installing new packages poetry would instead of copying, just hardlink or soflink/symlink files from this repository.

Besides the apparent gain of less disk usage *(we're talking many GB in many situations)*, I imagine that IDE:s that are indexing code would be more efficient if they were pointed to a single source, instead of needing to index multiple copies of the same file over and over.

## always-copy not utilized?
Anyhow, why I replied in this thread is that I tried to understand how this options was utilized, and to my surprise I found nowhere in the code where the setting `virtualenvs.options.always-copy` is actually utilized.

The setting is being picked up, and it's appearing in some unit tests, but it's never passed to any function.
So my take on it is that it's not utilized, but perhaps I'm overseeing something.

I searched these code bases:
- `python-poetry/poetry`
- `python-poetry/poetry-core`

Using the regex: `always.copy` *(case insensitive)*.

**Comment by dimbleby** [comment]
the option is simply passed through to virtualenv, see https://github.com/python-poetry/poetry/blob/081146956349e24d901c5187685df345e6621a98/src/poetry/utils/env/env_manager.py#L588
